zne/noise_amplification/folding_amplifier/glorious_folding_amplifier.py

```diff
@@ -0,0 +1,30 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Initial Function implmention. Will be evolved and documented more in depth"""
+from qiskit.circuit.library import Barrier
+from qiskit.dagcircuit import DAGCircuit
+
+
+def apply_folding(noisy_dag: DAGCircuit, original_dag: DAGCircuit, copies: int = 2) -> DAGCircuit:
+    """Initial function for now, will be evolved and documented more in depth"""
+
+    for _ in range(copies):
+        noisy_dag.compose(original_dag, inplace=True)
+        noisy_dag.apply_operation_back(Barrier(noisy_dag.num_qubits()), qargs=noisy_dag.qubits)
+    return noisy_dag
+
+
+def dag_duplicator(dag_circuit: DAGCircuit) -> DAGCircuit:
+    """Initial function for now, will be evolved and documented more in depth"""
+
+    return apply_folding(noisy_dag=dag_circuit.copy_empty_like(), original_dag=dag_circuit)
```

Comment:
Smart move using `dag_circuit.copy_empty_like()`. 

A better approach would be to remove `noisy_dag` from args and calling this inside the function body itself! Providing a `noisy_dag` is not necessary for what we are trying to do, otherwise using a pair of functions like this would be a very smart generalization. What you just did is known as a _facade_ in dev community (i.e. a function/class that simplifies a more general one). This is very useful indeed! But feels a bit like overegineering in this case.

Good job!!

Reply:
Removed it and restructured the function! 😁